File: backend/app/models/github.py
"""Pydantic models for GitHub App webhook activity and status."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class GitHubActivityItem(BaseModel):
    """Normalized GitHub webhook event stored for activity feeds."""

    id: str
    event_type: str
    action: str | None = None
    title: str
    body_preview: str | None = None
    repo_full_name: str | None = None
    actor_login: str | None = None
    html_url: str | None = None
    task_keys: list[str] = Field(default_factory=list)
    installation_id: str | None = None
    delivery_id: str | None = None
    received_at: datetime


class GitHubInstallationRecord(BaseModel):
    """Persisted (in-memory) GitHub App installation linked to Clarity."""

    installation_id: str
    account_login: str | None = None
    account_type: str | None = None
    repository_selection: str | None = None
    repositories: list[str] = Field(default_factory=list)
    suspended: bool = False
    team_id: UUID | None = None
    user_id: UUID | None = None
    connected_at: datetime
    updated_at: datetime


class GitHubConnectResponse(BaseModel):
    """Install URL for the Connect GitHub flow."""

    url: str
    state: str
    app_slug: str | None = None
    requires_session: bool = False
    detail: str | None = None


class GitHubCallbackResponse(BaseModel):
    """Result of the GitHub App setup redirect callback."""

    ok: bool
    installation_id: str | None = None
    linked: bool = False
    account_login: str | None = None
    detail: str | None = None


class GitHubStatusResponse(BaseModel):
    """Whether the GitHub App integration is configured and reachable."""

    configured: bool
    app_id_set: bool
    webhook_secret_set: bool
    private_key_loaded: bool
    installation_id_set: bool
    installation_reachable: bool | None = None
    connected: bool = False
    installation_id: str | None = None
    account_login: str | None = None
    account_type: str | None = None
    repositories: list[str] = Field(default_factory=list)
    repository_selection: str | None = None
    team_id: UUID | None = None
    user_id: UUID | None = None
    source: str | None = None
    detail: str | None = None

File: frontend/src/types/github.ts
/** GitHub activity types returned by `/api/v1/github/*`. */

export interface GitHubActivityItem {
  id: string;
  event_type: string;
  action: string | null;
  title: string;
  body_preview: string | null;
  repo_full_name: string | null;
  actor_login: string | null;
  html_url: string | null;
  task_keys: string[];
  installation_id: string | null;
  delivery_id: string | null;
  received_at: string;
}

export interface GitHubStatus {
  configured: boolean;
  app_id_set: boolean;
  webhook_secret_set: boolean;
  private_key_loaded: boolean;
  installation_id_set: boolean;
  installation_reachable: boolean | null;
  connected: boolean;
  installation_id: string | null;
  account_login: string | null;
  account_type: string | null;
  repositories: string[];
  repository_selection: string | null;
  team_id: string | null;
  user_id: string | null;
  source: string | null;
  detail: string | null;
}

export interface GitHubConnect {
  url: string;
  state: string;
  app_slug: string | null;
  requires_session: boolean;
  detail: string | null;
}
